### PR TITLE
Implement 1st tests

### PR DESCRIPTION
**Pull request by solarlodge**

### CODE REVIEW

File: src/main/java/de/cronn/diff/Main.java [hunk 1]
```diff
@@ -26,6 +28,8 @@
 
     private static String workingDir = FileHelper.getWorkingDir();
 
+	public static Long MAX_ALLOWED_FILESIZE_DIFFERENCE = 5000000l;
```
Comment (yudanaz) [inline src/main/java/de/cronn/diff/Main.java]: We should name the unit in the const name, i.e. `MAX_ALLOWED_FILESIZE_DIFFERENCE_IN_BYTES`.

File: src/main/java/de/cronn/diff/Main.java [hunk 2]
```diff
@@ -28,7 +28,7 @@
 
     private static String workingDir = FileHelper.getWorkingDir();
 
-	public static Long MAX_ALLOWED_FILESIZE_DIFFERENCE = 5000000l;
+	public static Long MAX_ALLOWED_FILESIZE_DIFFERENCE = 500000l;
```
Comment (yudanaz) [inline src/main/java/de/cronn/diff/Main.java]: This commit should be squashed into the previous commit (it even has the same commit message).

File: src/main/java/de/cronn/diff/CronnDiffToHtml.java [hunk 3]
```diff
@@ -9,7 +9,7 @@
 import de.cronn.diff.util.DiffToHtmlParameters;
 import de.cronn.diff.util.DiffToHtmlParameters.DiffType;
 
-public class CronnDiffToHtml {
+class CronnDiffToHtml {
```
Comment (yudanaz) [inline src/main/java/de/cronn/diff/CronnDiffToHtml.java]: Why?

File: src/main/java/de/cronn/diff/html/DirectoryDiffHtmlBuilder.java [hunk 4]
```diff
@@ -62,29 +62,29 @@
 
 	private static final String CSS_CLASS_ONE_LINE_HEADER = "one-line-header";
 
-	private String title;
+	private final String title;
```
Comment (yudanaz) [inline src/main/java/de/cronn/diff/html/DirectoryDiffHtmlBuilder.java]: I don't agree with adding these `final` keywords in this class. We might even want to re-assign one of these class members at some point, plus it blows up the code without need.

File: src/main/java/de/cronn/diff/html/FileDiffHtmlBuilder.java [hunk 5]
```diff
@@ -33,7 +33,7 @@ public FileDiffHtmlBuilder(DiffToHtmlParameters params) {
 		super(params);
 	}
 
-	private List<Tag> diffLines = new ArrayList<>();
+	private final List<Tag> diffLines = new ArrayList<>();
```
Comment (yudanaz) [inline src/main/java/de/cronn/diff/html/FileDiffHtmlBuilder.java]: no need for this final

File: src/main/java/de/cronn/diff/html/HtmlBuilder.java [hunk 6]
```diff
@@ -22,33 +22,33 @@
...
 
-	protected String outputDirForRelativePaths = null;
+	private String outputDirForRelativePaths = null;
 
 	protected abstract Tag createHead();
 
 	protected abstract Tag createBody();
 
-	public HtmlBuilder(DiffToHtmlParameters params) {
+	HtmlBuilder(DiffToHtmlParameters params) {
```
Comment (yudanaz) [inline src/main/java/de/cronn/diff/html/HtmlBuilder.java]: Why?

File: src/main/java/de/cronn/diff/html/HtmlBuilder.java [hunk 7]
```diff
@@ -22,33 +22,33 @@
 
 	private static final String CSS_FILE = "diffToHtml.css";
 
-	protected static final String EMPTY_LINE_HEIGHT = "16";
+	static final String EMPTY_LINE_HEIGHT = "16";
```
Comment (yudanaz) [inline src/main/java/de/cronn/diff/html/HtmlBuilder.java]: Why?

File: src/main/java/de/cronn/diff/html/HtmlBuilder.java [hunk 8]
```diff
@@ -22,33 +22,33 @@
...
 
-	protected static final String CSS_CLASS_LINE_NUMBER = "line-number";
+	static final String CSS_CLASS_LINE_NUMBER = "line-number";
```
Comment (yudanaz) [inline src/main/java/de/cronn/diff/html/HtmlBuilder.java]: Why? why? and below more why?s

Comment (yudanaz) [reply src/main/java/de/cronn/diff/html/HtmlBuilder.java]: Actually, if you want to change it, these constants could/should all be `public`.

File: src/main/java/de/cronn/diff/html/HtmlBuilder.java [hunk 9]
```diff
@@ -68,7 +68,7 @@ public String toString() {
 		}
 	}
 
-	protected String getRelativePath(String path) {
+	String getRelativePath(String path) {
```
Comment (yudanaz) [inline src/main/java/de/cronn/diff/html/HtmlBuilder.java]: why all this opening of these methods to package publicity? I don't get it...

File: src/main/java/de/cronn/diff/impl/java/JavaFileDiffToHtmlImpl.java [hunk 10]
```diff
@@ -19,21 +19,21 @@
...
+	private static final String IDENTICAL_FILES_PREFIX = "Files ";
 
-	public static final String IDENTICAL_FILES_SUFFIX = " are identical";
+	private static final String IDENTICAL_FILES_SUFFIX = " are identical";
 
-	public static final String IDENTICAL_LINE_SPLIT_STR = " and ";
+	private static final String IDENTICAL_LINE_SPLIT_STR = " and ";
 
-	protected DiffToHtmlParameters params;
+	final DiffToHtmlParameters params;
```
Comment (yudanaz) [inline src/main/java/de/cronn/diff/impl/java/JavaFileDiffToHtmlImpl.java]: you're relaxing access restrictions without need here again, don't.

File: src/main/java/de/cronn/diff/impl/java/JavaFileDiffToHtmlImpl.java [hunk 11]
```diff
@@ -92,23 +92,23 @@ private void setFileInfos(FileDiffHtmlBuilder htmlBuilder, DiffToHtmlParameters
 		setRightFileInfo(htmlBuilder, params);
 	}
 
-	protected void setLeftFileInfo(FileDiffHtmlBuilder htmlBuilder, DiffToHtmlParameters params) throws IOException {
+	void setLeftFileInfo(FileDiffHtmlBuilder htmlBuilder, DiffToHtmlParameters params) throws IOException {
```
Comment (yudanaz) [inline src/main/java/de/cronn/diff/impl/java/JavaFileDiffToHtmlImpl.java]: why?

File: src/test/java/de/cronn/diff/MainTestBase.java [hunk 12]
```diff
@@ -16,22 +16,22 @@
...
-	protected static final String INPUT_TEXT_WITH_SPACES_2 = TEST_DATA_INPUT_DIR + "textWithSpaces2";
-	protected static final String INPUT_TEXT_WITHOUT_SPACES = TEST_DATA_INPUT_DIR + "textWithoutSpaces";
-	protected static final String INPUT_DIR_WITH_SPACES_1 = TEST_DATA_INPUT_DIR + "dirWithSpaces1";
-	protected static final String INPUT_DIR_WITH_SPACES_2 = TEST_DATA_INPUT_DIR + "dirWithSpaces2";
-	protected static final String INPUT_DIR_WITHOUT_SPACES = TEST_DATA_INPUT_DIR + "dirWithoutSpaces";
-	protected static final String INPUT_TEXT_1_1 = TEST_DATA_INPUT_DIR + "text1_1.example";
-	protected static final String INPUT_TEXT_1_2 = TEST_DATA_INPUT_DIR + "text1_2.example";
-	protected static final String SMALL_FILE = TEST_DATA_INPUT_DIR + "smallfile";
-	protected static final String BIG_FILE = TEST_DATA_INPUT_DIR + "bigfile";
+	static final String INPUT_DIR_1 = TEST_DATA_INPUT_DIR + "dir1_1";
```
Comment (yudanaz) [inline src/test/java/de/cronn/diff/MainTestBase.java]: WH-H-HY???

File: src/test/java/de/cronn/diff/TestBase.java [hunk 13]
```diff
@@ -22,23 +22,23 @@
...
-	public static final String HTML_SUFFIX = ".html";
+	private static final String HTML_SUFFIX = ".html";
 
 	private static final String SYS_OUT_SUFFIX = ".sysouterr";
 
 	private static boolean outputDirChecked = false;
 
 	@Rule
-	public TestName testName = new TestName();
+	public final TestName testName = new TestName();
```
Comment (yudanaz) [inline src/test/java/de/cronn/diff/TestBase.java]: unnecessary final

File: data/test/input/not_so_empty_output/here_lies_a_file [hunk 14]
```diff
@@ -0,0 +1 @@
+Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
```
[the remark this answers is not in the document]
Comment (solarlodge) [reply data/test/input/not_so_empty_output/here_lies_a_file]: Honestly - I dunno. Maybe some leftover...

File: src/main/java/de/cronn/diff/Main.java (hunk 2 above)
Comment (solarlodge) [reply src/main/java/de/cronn/diff/Main.java]: done.

File: src/main/java/de/cronn/diff/impl/java/JavaDirDiffToHtmlImpl.java [hunk 15]
```diff
@@ -23,10 +24,12 @@
 import de.cronn.diff.util.FileHelper;
 
 public class JavaDirDiffToHtmlImpl extends JavaFileDiffToHtmlImpl {
-	
+
 	private static final String UNIQUE_FILE_PREFIX = "Only in ";
 
 	private static final String UNIQUE_LINE_SPLIT_STR = ": ";
+	public static final String tooManyDifferencesErrorMsg = "One of the directories seems to be empty or the difference in diffing is too big. Abort!";
```
[the remark this answers is not in the document]
Comment (solarlodge) [reply src/main/java/de/cronn/diff/impl/java/JavaDirDiffToHtmlImpl.java]: done.

File: src/main/java/de/cronn/diff/Main.java (hunk 1 above)
Comment (solarlodge) [reply src/main/java/de/cronn/diff/Main.java]: done.

File: src/main/java/de/cronn/diff/impl/java/JavaDirDiffToHtmlImpl.java [hunk 16]
```diff
@@ -37,19 +40,41 @@ public DiffToHtmlResult runDiffToHtml() throws IOException {
...
+				smallList = rightSortedFilesAndDirs.size();
+				bigList = leftSortedFilesAndDirs.size();
+			} else if (rightSortedFilesAndDirs.size() > leftSortedFilesAndDirs.size()) {
+				smallList = leftSortedFilesAndDirs.size();
+				bigList = rightSortedFilesAndDirs.size();
+			}
+			int halfList = bigList/2;
+			status = halfList <= smallList;
+		}
+		return status;
```
[the remark this answers is not in the document]
Comment (solarlodge) [reply src/main/java/de/cronn/diff/impl/java/JavaDirDiffToHtmlImpl.java]: whatever...

**Pull request merged**